# Zero to Hero

## 1.09 Dividir en Training y Testing
## version libreria caret

El objetivo es hacer una division del dataset en training/testing que sea estratificada en la clase
<br>
https://es.wikipedia.org/wiki/Muestreo_estratificado

In [1]:
rm( list=ls())
gc()

library( "data.table")   #cargo la libreria  data.table
library( "rpart")  #cargo la libreria  rpart
library( "rpart.plot")

options(repr.plot.width=20, repr.plot.height=10) 
setwd("C:\\Users\\hmreu\\Documents\\Especialización en Ciencias de Datos\\Mineria de Datos\\")  #Aqui se debe poner la ruta de la PC local

used    (Mb) gc trigger (Mb) max used (Mb)
Ncells  594960 31.8 1267284    67.7 1267284  67.7
Vcells 1065189  8.2 8388608    64.0 1603908  12.3

In [2]:
dataset <- data.table::fread("./datasets/paquete_premium_202011.csv")   #cargo el dataset

In [3]:
library("caret")

Loading required package: lattice

Loading required package: ggplot2

Registered S3 methods overwritten by 'tibble':
  method     from  
  format.tbl pillar
  print.tbl  pillar



https://www.rdocumentation.org/packages/caret/versions/6.0-88/topics/createDataPartition
<br>
la funcion  **createDataPartition** devolverá un vector de posiciones, las que cumplen con la particion indicada
<br>
*p= 0.5* significa que queremos el 50% de los registros
<br>
*dataset$clase_ternaria*   es el vector con la clase ternaria, que es por donde se estratificará
<br>

In [4]:
train_rows <- createDataPartition(dataset$clase_ternaria, p= 0.50, list= FALSE)

veamos que ha quedado en  los primeros registros del vector **train_rows**

In [5]:
head( train_rows )

Resample1
[1,]  2       
[2,]  3       
[3,]  4       
[4,]  7       
[5,]  9       
[6,] 11

Aqui ya vemos que la particion tiene un suconjunto de los registros

ahora creamos los datasets de training y testing

In [6]:
dtrain <- dataset[ train_rows]

-train_rows  significa el complemento  (no confundir con numeros negativos)

In [7]:
dtest <- dataset[ -train_rows]

Compruebo la division

In [8]:
nrow( dtrain)
nrow( dtest )
nrow(dtrain) + nrow(dtest)
nrow( dataset)

[1] 79695

[1] 79693

[1] 159388

[1] 159388

Compruebo que la particion es estratificada

In [9]:
dataset[  , .N, clase_ternaria]
dtrain[  , .N, clase_ternaria]
dtest[  , .N, clase_ternaria]

clase_ternaria N     
1 CONTINUA       158229
2 BAJA+1            511
3 BAJA+2            648

clase_ternaria N    
1 CONTINUA       79115
2 BAJA+1           256
3 BAJA+2           324

clase_ternaria N    
1 CONTINUA       79114
2 BAJA+1           255
3 BAJA+2           324

La division se ha construido en forma estratificada

La cantidad de BAJA+2 es 648,  y se observa que training tiene 324 BAJA+2 ,exactamente la mitad de 648,  y testing tambien 324 BAJA+2